### PR TITLE
Added `reindex` method in ``GenericTimeSeries``

### LINKED ISSUE #5369: Add `reindex` method to `GenericTimeseries`
Currently if I want to re-index some data the code is messy. In this case, I am re-indexing `mag_data` on to the timestamps of `swa_data`, both of which are `GenericTimeseries` objects:

```python
mag_df = mag_data.to_dataframe()
mag_reindexed = mag_data.to_dataframe().reindex(swa_data.index,
                                                method='nearest')
mag_reindexed = GenericTimeSeries(mag_reindexed, units=mag_data.units)
```

It would be good to add a `reindex` method to `GenericTimeseries` to wrap this functionality, so instead I could just do:

```python
mag_reindexed = mag_data.reindex(saw_data.index)
```

### PR DESCRIPTION
**Pull request by devanshshukla99**
### Description

Fixes #5369

### CODE REVIEW

File: sunpy/timeseries/timeseriesbase.py [hunk 1]
```diff
@@ -445,6 +445,26 @@ def concatenate(self, others, same_source=False, **kwargs):
...
+        Returns
+        -------
+        `~sunpy.timeseries.TimeSeries`
+            A `~sunpy.timeseries.TimeSeries` with new index.
+
+        Notes
+        -----
+        This method is a wrapper around `~pandas.DataFrame.reindex`
+        """
+        return GenericTimeSeries(self._data.reindex(index, **kwargs), units=self.units)
```
Comment (nabobalis) [inline sunpy/timeseries/timeseriesbase.py:466]: Is the timeseries metadata taken account of?

Comment (devanshshukla99) [reply sunpy/timeseries/timeseriesbase.py:466]: is copying metadata enough or should I modify it?

Comment (nabobalis) [reply sunpy/timeseries/timeseriesbase.py:466]: Copying it should be enough? I'm not too familiar with time series metadata. I don't think there is a time value but I could be quite wrong.

Comment (devanshshukla99) [reply sunpy/timeseries/timeseriesbase.py:466]: look like there is,
attached metadata for goes xrs test data:
```console
[(   <sunpy.time.timerange.TimeRange object at 0x7f36b841ef10>
      Start: 2011-06-06 23:59:59
      End:   2011-06-07 23:59:57
      Center:2011-06-07 11:59:58
      Duration:0.9999730324069096 days or
             23.99935277776583 hours or
             1439.9611666659498 minutes or
             86397.66999995698 seconds,
  ['xrsa', 'xrsb'],
  MetaDict([('simple', True),
            ('bitpix', 8),
            ('naxis', 0),
            ('extend', True),
            ('date', '26/06/2012'),
            ('numext', 3),
            ('telescop', 'GOES 15'),
            ('instrume', 'X-ray Detector'),
            ('object', 'Sun'),
            ('origin', 'SDAC/GSFC'),
            ('date-obs', '07/06/2011'),
            ('time-obs', '00:00:00.000'),
            ('date-end', '07/06/2011'),
            ('time-end', '23:59:57.632'),
            ('comment',
             'Energy band information given in extension 1Fluxes information given in extension 2Status word information given in extension 3Ref.: Solar X-Ray Measurements from SMS-1, SMS-2, and GOES-1;Information for Data Users. Donnelly et al,  June 1977.NOAA TM ERL SEL-48Ref.: SMS GOES Space Environment Monitor Subsystem,Grubb, Dec 75, NOAA, Technical Memorandum ERL SEL-42.Ref.:Expresions to Determine Temperatures and EmissionMeasures for Solar X-ray events from GOES Measurements.Thomas et al, 1985, Solar Physics 95, pp 323-329.'),
            ('history', ''),
            ('keycomments',
             {'SIMPLE': 'Written by IDL:  Tue Jun 26 09:32:13 2012',
              'EXTEND': 'File contains extensions',
              'DATE': 'file creation date',
              'NUMEXT': 'no. of extensions in file',
              'TELESCOP': 'spacecraft',
              'ORIGIN': 'written by Solar DAC at GSFC',
              'DATE-OBS': 'date of first obs.(DD/MM/YYYY)',
              'TIME-OBS': 'UT time of first obs.(HH:MM:SS.XXX)',
              'DATE-END': 'date of last obs.(DD/MM/YYYY)',
              'TIME-END': 'UT time of last obs.(HH:MM:SS.XXX)'})]))]
```

Comment (nabobalis) [reply sunpy/timeseries/timeseriesbase.py:466]: Copy it and then others can chime in if anything needs updating after.

Comment (devanshshukla99) [reply sunpy/timeseries/timeseriesbase.py]: time range issue should be fixed now👍

File: changelog/5373.feature.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Added :meth:``sunpy.timeseries.GenericTimeSeries.reindex`` in ``sunpy.timeseries.GenericTimeSeries`` as a wrapper for ``~pandas.DataFrame.reindex``.
```
Comment (nabobalis) [inline changelog/5373.feature.rst:1]:
```suggestion
Added :meth:`sunpy.timeseries.GenericTimeSeries.reindex` in `sunpy.timeseries.GenericTimeSeries` as a wrapper for `~pandas.DataFrame.reindex`.
```

**Comment by dstansby** [comment]
Thanks for the PR. Reading over the additions, I'm wondering if `reindex` should just take another `GenericTimeSeries` object instead of a `pandas.DateTimeIndex`, so the sunpy layer remains independent of the implementation (which is pandas at the moment). I think we should discuss this on a community call, so I'll post an update on what we decide.

**Comment by dstansby** [comment]
Just discussed this at the weekly call - we decided that `index` should be accepted as both a `pandas.TimeSeriesIndex` and a `GenericTimeSeries`. I'm happy to approve this as is and add support for `GenericTimeSeries` in another PR, or you can add that support in this PR - let us know.

**Comment by devanshshukla99** [comment]
> Just discussed this at the weekly call - we decided that `index` should be accepted as both a `pandas.TimeSeriesIndex` and a `GenericTimeSeries`. I'm happy to approve this as is and add support for `GenericTimeSeries` in another PR, or you can add that support in this PR - let us know.

should be rectified now👍

File: sunpy/timeseries/tests/test_timeseriesbase.py [hunk 3]
```diff
@@ -906,6 +906,18 @@ def test_ts_sort_index(generic_ts):
     assert generic_ts.sort_index().to_dataframe().equals(generic_ts.to_dataframe().sort_index())
 
 
+def test_ts_reindex(generic_ts):
+    dates = generic_ts.time_range.start + TimeDelta(np.arange(60)*u.minute)
+    new_index = dates.isot.astype('datetime64')
+    # Test for pandas.DatetimeIndex as index
+    generic_ts_reindexed_1 = generic_ts.reindex(new_index, method="nearest")
+    df_selected = generic_ts.to_dataframe().loc[new_index]
+    assert generic_ts_reindexed_1.to_dataframe().equals(df_selected) is True
```
Comment (dstansby) [inline sunpy/timeseries/tests/test_timeseriesbase.py:915]:
```suggestion
    assert generic_ts_reindexed_1.to_dataframe().equals(df_selected)
```

Comment (dstansby) [reply sunpy/timeseries/tests/test_timeseriesbase.py:915]: `is True` doesn't add anything here

File: sunpy/timeseries/tests/test_timeseriesbase.py [hunk 4]
```diff
@@ -906,6 +906,18 @@ def test_ts_sort_index(generic_ts):
...
+    # Test for sunpy.timeseries.TimeSeries as index
+    generic_ts_reindexed_2 = generic_ts.reindex(generic_ts_reindexed_1, method="nearest")
+    assert generic_ts_reindexed_2.to_dataframe().equals(generic_ts_reindexed_1.to_dataframe()) is True
```
Comment (dstansby) [inline sunpy/timeseries/tests/test_timeseriesbase.py:918]:
```suggestion
    assert generic_ts_reindexed_2.to_dataframe().equals(generic_ts_reindexed_1.to_dataframe())
```

File: sunpy/timeseries/timeseriesbase.py [hunk 5]
```diff
@@ -445,6 +445,33 @@ def concatenate(self, others, same_source=False, **kwargs):
...
+            Another `~sunpy.timeseries.TimeSeries` or a valid index column.
+
+        Returns
+        -------
+        `~sunpy.timeseries.TimeSeries`
+            A `~sunpy.timeseries.TimeSeries` with new index.
+
+        Notes
+        -----
+        This method is a wrapper around `~pandas.DataFrame.reindex`
```
Comment (dstansby) [inline sunpy/timeseries/timeseriesbase.py:464]:
```suggestion
        This method is a wrapper around `pandas.DataFrame.reindex`; all additional
        keyword arguments are passed to this method.

File: sunpy/timeseries/timeseriesbase.py [hunk 6]
```diff
@@ -445,6 +445,33 @@ def concatenate(self, others, same_source=False, **kwargs):
         object._sanitize_units()
         return object
 
+    def reindex(self, index, **kwargs):
+        """
+        Returns a new time series with the chosen index.
```
Comment (dstansby) [inline sunpy/timeseries/timeseriesbase.py:450]:
```suggestion
        Returns a new time series with a new index.
        
        By default values on the new time series are filled using a
        nearest valid observation method. See `~pandas.DataFrame.reindex`
        for the different re-indexing options available.
```

**Review by dstansby — APPROVED** [review]
Minor suggestions to improve docs, but looks 👍 overall

File: sunpy/timeseries/timeseriesbase.py (hunk 6 above)
Comment (devanshshukla99) [reply sunpy/timeseries/timeseriesbase.py]: I believe the default method for `reindex` is `None`, shall I add another param to default it to `nearest`?

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Thanks again @devanshshukla99

**Pull request merged**